homeassistant/components/stream/worker.py

```diff
@@ -154,43 +152,42 @@ def check_flush_part(self, packet: av.Packet) -> None:
                 stream_id=self._stream_id,
                 init=self._memory_file.getvalue(),
             )
-            self._segment_last_write_pos = byte_position
+            self._memory_file_pos = self._memory_file.tell()
+            self._part_start_dts = self._segment_start_dts
             # Fetch the latest StreamOutputs, which may have changed since the
             # worker started.
             for stream_output in self._outputs_callback().values():
                 stream_output.put(self._segment)
         else:  # These are the ends of the part segments
-            self._segment.parts.append(
-                Part(
-                    duration=float(
-                        (packet.dts - self._part_start_dts) * packet.time_base
-                    ),
-                    has_keyframe=self._part_has_keyframe,
-                    data=self._memory_file.getbuffer()[
-                        self._segment_last_write_pos : byte_position
-                    ].tobytes(),
-                )
-            )
-            self._segment_last_write_pos = byte_position
-            self._part_start_dts = packet.dts
-            self._part_has_keyframe = False
+            self.flush_part(packet, last_part=False)
 
-    def flush(self, duration: Fraction, packet: av.Packet) -> None:
-        """Create a segment from the buffered packets and write to output."""
-        self._av_output.close()
+    def flush_part(self, packet: av.Packet, last_part: bool) -> None:
+        """Output a part from the most recent bytes in the memory_file."""
         assert self._segment
-        self._segment.duration = float(duration)
-        # Also flush the part segment (need to close the output above before this)
+        self._memory_file.seek(self._memory_file_pos)
         self._segment.parts.append(
             Part(
                 duration=float((packet.dts - self._part_start_dts) * packet.time_base),
                 has_keyframe=self._part_has_keyframe,
-                data=self._memory_file.getbuffer()[
-                    self._segment_last_write_pos :
-                ].tobytes(),
+                data=self._memory_file.read(),
             )
         )
+        if last_part:
+            self._memory_file_pos = 0
+        else:
+            self._memory_file_pos = self._memory_file.tell()
+            self._part_start_dts = packet.dts
+        self._part_has_keyframe = False
+
+    def flush(self, duration: Fraction, packet: av.Packet) -> None:
+        """Close the segment and give it a duration, making it complete."""
+        # Closing the av_output will write the remaining buffered data to the
+        # memory_file as a new moof/mdat.
+        self._av_output.close()
+        self.flush_part(packet, last_part=True)
         self._memory_file.close()  # We don't need the BytesIO object anymore
+        assert self._segment
+        self._segment.duration = float(duration)
```

Comment:
There is an issue to consider here. Going forward we will add a `part_put` to signal that a `Part` is ready, just like `put` signals that a `Segment` is ready. It makes sense to put that call in `flush_part`.
However, the `Part` itself doesn't store any information of whether it is the last `Part` or not - we currently rely on the `Segment.complete` property (whether `Segment` has a non zero `duration`). But we want this information to be available when `part_put` is called. To satisfy this we can either:
1) Move this `self._segment.duration` assignment up into `flush_part`
2) Add a `last_part` boolean member to `Part`

Reply:
Thanks for calling this out. I don't have a strong opinion yet. Another angle I was thinking about was having methods on `Segment` like `put_part` but i'm not sure its better.  I think there used to be logic in `Segment` to track part state, but then it seemed simpler to move it here, so not sure that is better.

Reply:
Hmm, not sure. I suppose having a last_part boolean is a bit cleaner, but then we have a little bit of duplicated information (although I don't think there's a consistency problem with that duplicated data)